### ISSUE TITLE
deferred: decide public Python binding after Rust facade stabilization

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/TeNeT#1086]
## Status

The Rust-side physical expansion work is internal and the public Python/JAX binding is intentionally deferred. The existing Python scaffold is not a supported expansion API.

## Deferred decision

Re-audit the desired user surface after the Rust facade and documentation reconciliation in #1112. Any future binding must expose the stable provider-typed TeNeT API without leaking internal expansion plans, opaque `SectorId` values, or backend ownership details.

This issue does not authorize a new `PhysicalExpansionPlan` public API or JAX-specific tensor abstraction. If Python work resumes, define a separate design issue with current source/tests as evidence.

**Comment by Ryo-wtnb11** [comment]
Initial source audit (2026-08-12): the repository already has the core capability needed for this issue. `tenet-sectors` defines `PhysicalFusionBasis` for checked carrier dimensions and local fusion-tensor elements; SU(2) implements it through the Racah authority. `tenet-tensors::physical` already stages carrier dimensions, local tensors, dual maps, tree embeddings, and reduced-block layout, then exposes `expand_physical_host`. The missing boundary is an immutable, provider-generic expansion-plan value that owns the completed structural map and provenance, followed by a narrow Python transport layer. The first implementation should reuse this path rather than duplicate CGC/F-symbol formulas or introduce a second cache. Before coding, the design must specify plan identity, real/complex scalar transport, typed unsupported-capability errors, and an independent SU(2) oracle.

**Comment by Ryo-wtnb11** [comment]
Implementation scope decision (2026-08-12): the first PR will establish the provider-generic Rust expansion-plan boundary and contract tests, reusing `PhysicalFusionBasis` and the existing staged physical expansion path. It will not add a new CGC/F/R implementation, duplicate a coefficient cache, or introduce a speculative Python dependency. Python transport will consume the stable Rust plan in a follow-up once the ownership, identity, scalar, and failure semantics are proven.

**Comment by Ryo-wtnb11** [comment]
Implementation audit (2026-08-12): the Python binding cannot safely start from the existing API. `FusionRule` is a concrete Rust-provider trait rather than a Python-held object-safe boundary, and hom-space APIs are parameterized by concrete `R`/`Arc<R>`. Existing Python support is planner-only (cotengra subprocess). A direct PyO3 wrapper would duplicate provider identity/coefficient authority or introduce unsafe callbacks.\n\nThe prerequisite is now split to a Rust-only provider adapter boundary issue. #1086 remains the end-to-end contract; after the adapter is implemented and independently tested, #1097 can add the separate publish=false wheel crate.

**Comment by Ryo-wtnb11** [comment]
Prerequisite split: #1101 (closed provider adapter boundary). This keeps #1086 as the end-to-end Python contract while making the Rust/provider ownership boundary independently reviewable.

**Comment by Ryo-wtnb11** [comment]
Follow-up audit (2026-08-12): the initial PyO3 scaffold was intentionally not merged. It only exposed SU(2) sector metadata (`dimension`, `dual`, `fusion_channels`) and did not implement the required immutable expansion plan, real/complex application, provenance metadata, typed capability failures, or independent oracle tests. It was also not reproducible in the current environment without the PyO3 ABI3 compatibility override. #1086 remains open; the next implementation must target the actual `expansion_plan().apply()` contract rather than a provider metadata facade.

**Comment by Ryo-wtnb11** [comment]
## Provider-generic implementation plan (design gate)

The public contract remains provider-generic; SU(2) is only the first capability and oracle.

### Commit-sized phases

1. **Rust contract types** — define provider-neutral plan metadata, capability identity, layout/provenance fields, and typed errors. No Python dependency and no provider-specific formulas.
2. **Generic Rust plan façade** — adapt the existing `PhysicalExpansionPlan` behind one provider-generic entry point. Unsupported providers fail before a plan is returned; exact provider allocation and structural layout remain checked by the existing implementation.
3. **SU(2) capability** — register the first concrete provider capability and add multi-rank, multi-sector, dual-orientation, real, and complex Rust contract tests.
4. **Independent oracle** — verify unit-coordinate reconstruction, linearity, and invariant-space correctness against an oracle that does not call the expansion implementation.
5. **Separate Python wheel scaffold** — add a `publish = false` binding crate only after the Rust contract is stable. The Python API must expose one provider-generic plan/apply entry point, not `*_su2` methods.
6. **NumPy transport and failure mapping** — owned Fortran-contiguous output, f64/complex128 inputs, typed shape/dtype/provenance/capability errors, and clean-environment reproducibility.

### Explicit boundary

The already merged SU(2) methods in #1102 are a prerequisite convenience layer, not the #1086 public contract. They must not become the Python API. Later providers should implement the same capability without changing the Python surface.

No phase is complete until its tests and documentation are committed separately; do not squash unrelated phases.

**Comment by Ryo-wtnb11** [comment]
Phase A merged as PR #1103. `PhysicalExpansionPlan::provider_identity()` now exposes the existing provider-generic `RuleIdentity` without adding a second cache or representation. All CI checks passed. Stable convention fingerprint and canonical semantic plan ID remain explicit design gates for the next phase; no invented fingerprint was added.

**Comment by Ryo-wtnb11** [comment]
## Design decision: descriptor before digest (2026-08-12)

Adopt the proposed separation. Do not introduce a hashed `semantic_plan_id` first. The next phase will define a canonical descriptor and its equality/serialization contract.

### Identity layers

- **Execution provenance:** the plan-owned provider allocation and existing `Arc::ptr_eq` checks.
- **In-process semantic identity:** existing `RuleIdentity`; useful for Rust equality but not a stable cross-process identifier.
- **Python metadata and reproducibility:** a canonical `ExpansionPlanDescriptorV1`.
- **Optional short display/file name:** a digest derived from the descriptor later; never the correctness authority.

### Descriptor contract

The design phase must specify `PhysicalBasisConventionId` and `ExpansionPlanDescriptorV1` without inventing a second coefficient cache. The descriptor must cover the convention epoch/payload, ordered codomain and domain legs, sectors and degeneracies, dual flags, ordered fusion-tree keys, reduced-coordinate/block ordering, physical shape/axis order, scalar convention, and schema version.

The provider convention identity is capability-specific, not the whole `RuleIdentity`. A provider without an authoritative stable convention identity must return a typed error when a plan is exported for Python or serialization.

### Required tests

- Equivalent provider allocations produce equal descriptors, while execution with a tensor from a distinct allocation remains rejected.
- Changing sector order, degeneracy, dual flags, tree order, reduced-coordinate order, axis layout, or convention epoch changes the descriptor.
- Rebuilding the same plan produces identical canonical bytes.
- No hidden global catalog or new plan cache is introduced in this phase.

TensorKit's FusionTree/HomSpace semantic components support this structural direction, but the versioned stable convention identity is a TeNeT contract and must be explicit.

**Comment by Ryo-wtnb11** [comment]
Design gate split to #1104: define the canonical physical expansion descriptor and provider convention identity before adding Python serialization or a digest. Existing `RuleIdentity` remains in-process only; exact provider `Arc` remains execution provenance. #1086 stays provider-generic and open.

**Comment by Ryo-wtnb11** [comment]
## API decision: provider-bound spaces, provider-generic plans (2026-08-12)

The Python API is provider-generic after space construction.

```python
su2 = tenet.SU2()
hom = tenet.HomSpace(codomain=(...), domain=(...))

plan = hom.expansion_plan()
explicit = plan.apply(reduced)
```

Provider-specific code is limited to constructing provider-bound spaces. The plan API is shared by every provider that supports carrier-basis expansion.

The first common plan surface is:

- `dense_shape: tuple[int, ...]`;
- `reduced_size: int`;
- read-only `dense_indices: np.ndarray[int64]`;
- read-only `reduced_indices: np.ndarray[int64]`;
- read-only `coefficients: np.ndarray` (`float64` or `complex128`);
- `semantic_key`: canonical descriptor bytes;
- immutable `metadata` containing layout and convention information;
- `apply(reduced)` returning an owned dense result.

The COO view is also the JAX integration boundary. TeNeT does not add `to_jax()` or a JAX-specific binding:

```python
flat = jnp.zeros(math.prod(plan.dense_shape), dtype=...)
flat = flat.at[plan.dense_indices].add(
    plan.coefficients * alpha[plan.reduced_indices]
)
tensor = flat.reshape(plan.dense_shape)
```

Rust should extend the existing `PhysicalExpansionPlan` minimally with a descriptor, a read-only COO view, and common apply semantics. It must not add `build_su2_expansion_plan()`, a `provider="su2"` plan entry point, a provider-specific plan type, or a JAX adapter. The provider is derived from the `HomSpace`.

This matches the existing five-field JAX exporter contract and allows replacing the TensorKit exporter with the TeNeT plan exporter without a framework-specific TeNeT API.

**Comment by Ryo-wtnb11** [comment]
Redesign required (2026-08-12): `expand_physical_host`, `PhysicalExpansionPlan`, and the COO view are internal Rust mechanisms, not necessarily the public Python API. PR #1105 is paused/closed pending a minimal user-facing design. The next design audit will start from the user operation (calling TeNeT from Python) and compare the existing typed facade with tenferro-rs/symmray, rather than exposing internal expansion machinery by default.

**Comment by Ryo-wtnb11** [comment]
## Binding architecture audit (2026-08-12)

The comparison changed the design direction.

- `tenferro-rs` currently has no PyO3/maturin Python binding. Its repository contains a **planned** `tenferro-capi` design: opaque tensor handles, status codes, f64-first functions, and DLPack import/export. It is not an existing Python tensor API we can simply reuse.
- `symmray` is Python-native: its array object directly retains indices, charge/symmetry, and block data, and presents ndarray-like operations. It does not require users to construct an explicit expansion-plan object for ordinary use.

Therefore TeNeT should not expose `expand_physical_host`, `PhysicalExpansionPlan`, COO views, or semantic IDs as the ordinary Python API. Those remain internal Rust mechanisms unless a measured downstream use requires them.

The next design target is a thin binding around the existing typed tensor/space façade, with provider-specific code limited to constructing a provider-bound space. Ordinary operations should be called on the tensor/space object. If a language-neutral boundary is needed later, follow the tenferro direction (opaque handle + explicit status/ownership, possibly DLPack), rather than inventing a large TeNeT-specific plan protocol.

The earlier plan/COO PR #1105 was intentionally closed pending this redesign. #1104's descriptor work is deferred until a real serialization or cross-process consumer exists.

**Comment by Ryo-wtnb11** [comment]
## Reset design direction (2026-08-12)

The code added during the earlier #1086 exploration remains in Git history and is not being deleted or reverted. However, it must not constrain the next design.

In particular, `PhysicalExpansionPlan`, `expand_physical_host`, `provider_identity`, and the closed SU(2) facade methods are experimental/internal implementation candidates from the exploration phase, not the settled Python API. We will restart the API design from the user operation: call TeNeT from Python and obtain/use a tensor value with minimal ceremony.

The next design pass will compare the existing typed facade with tenferro-rs and symmray, decide whether any plan object is needed at all, and only then choose which existing Rust code to retain, wrap, or replace internally. No public descriptor, COO protocol, semantic ID, JAX adapter, or provider-specific plan entry point is assumed.

**Comment by Ryo-wtnb11** [comment]
## Binding-first redesign (2026-08-12)

The scope is reset from carrier-basis expansion to the basic Python binding itself.

A Python binding means that Python calls TeNeT's tensor API and TeNeT executes the symmetric-tensor operation in Rust. It is not the same as casting a TeNeT value to NumPy/JAX. `to_dense`, JAX PyTrees, expansion maps, COO transport, and `PhysicalExpansionPlan` are separate interoperability work and are not part of the first binding milestone.

### First milestone

Expose a small provider-generic Python facade around the existing everyday Rust API:

```python
import tenet

space = tenet.Space.u1([(-1, 1), (0, 2), (1, 1)])
other = space.dual()
tensor = tenet.Tensor.zeros(dtype="f64", codomain=[space], domain=[other])
result = tensor.permute(...)
```

The exact spelling may change, but the first public surface should cover only:

- provider-bound `Space` construction and `dual`;
- tensor construction (`zeros`, then `random`);
- read-only shape/dtype/space metadata;
- one or two existing basic tensor operations with typed error mapping.

No provider-specific Python operation such as `build_su2_*`, and no plan/COO/semantic-ID object.

`autoray` is not the binding mechanism. It may later reduce duplicated NumPy/JAX/PyTorch array dispatch in an optional interoperability layer, but PyO3/maturin (or an equivalent FFI boundary) is required for Python to call the Rust facade.

The existing PyO3 scaffold was not merged because it exposed only SU(2) metadata and did not constitute a usable TeNeT binding. The next PR must provide a reproducible `publish = false` binding crate and a smoke test that imports the module and executes the common API. Dense/JAX conversion remains a later, separately scoped issue.

**Comment by Ryo-wtnb11** [comment]
## Correction on product-space exposure

`Space::fz2` already exists in the Rust facade and is a safe first Python binding candidate.

`Space::product` and `Space::fz2_u1_su2` also exist, but they are not a general product-symmetry API. They are concrete constructors for the currently implemented `U1 × fZ2` and `(fZ2 × U1) × SU2` encodings, including bounded TensorKit-compatible sector-ID encoding and a fallible capacity check.

Therefore the Python binding must not present `product(...)` as if it accepted arbitrary providers or arbitrary product categories. The first binding milestone should expose `fz2` together with the existing elementary providers. Product constructors should remain deferred until the provider/product abstraction and Python representation are explicitly designed. If exposed later, their names and documentation must state the concrete category and return typed encoding errors.

The corrected order is:

1. existing elementary provider spaces (`u1`, `z2`, `fz2`, `su2`);
2. tensor construction and metadata;
3. `contract` and `permute`;
4. compact decompositions;
5. a separately designed product-symmetry binding;
6. truncation/eigenvalue-rich APIs.

**Comment by Ryo-wtnb11** [comment]
Correction: tenet-core already has generic ProductFusionRule/ProductFusionRuleExt::product. The issue is the user facade closed UserRuleContext variants and concrete Space constructors, not the core product machinery. Generic facade work is split to #1107; #1086 will consume the composed-provider path rather than expose fixed U1FZ2/FZ2U1SU2 Python methods.

**Comment by Ryo-wtnb11** [comment]
Rust-first decision (2026-08-13): Python/JAX binding and expansion export are deferred. We will first stabilize the TeNeT Rust facade, provider composition, contraction/factorization, and backend boundaries. No Python API is part of the current milestone.

**Comment by Ryo-wtnb11** [comment]
Scope correction after current-main audit: Rust physical expansion internals landed in #1096 and related provider-generic work, but the public Python/JAX binding remains intentionally deferred. This issue should not be read as a complete Python API milestone. Python export/binding is deferred until the Rust facade and ownership boundary are re-audited under #1112; no new expansion-plan public API is implied.

**Comment by Ryo-wtnb11** [comment]
Reviewed after the 0.1.1 registry publication and documentation reconciliation. This remains deferred/design-only: no implementation is implied by the release. Keep open until the separate design gate is intentionally resumed.

**Comment by Ryo-wtnb11** [comment]
Decision: Python binding/py-wheel/serialization work is intentionally not part of TeNeT. Keep Python-specific APIs and expansion descriptors out of the Rust public surface. Closing as out of scope; reopen only if the project explicitly adopts a Python deliverable.
